Feature/3399 default store redirect

rain2o opened the pull request:
### Related Issues
<!--  Put related issue number which this PR is closing. For example #123 -->

closes #3399 

### Short Description and Why It's Useful
<!-- describe in a few words what is this Pull Request changing and why it's useful -->

Using @cewald recommendation, and @pkarw  suggestion to submit it as a `src` module. I made a few modifications to the original version provided by @cewald.

Added a new config field `appendDefaultStoreCode`, which is checked before executing this module's logic.
If the conditions are met, it will attempt to redirect any requests without a store code to the `defaultStoreCode` value. If there is no value provided in `defaultStoreCode` it will use the first store code found in `mapStoreUrlsFor`.

Before redirecting, it also executes the hook `beforeRedirectToDefaultStore` passing the `storeCode` parameter. This allows 3rd-party modules to use this hook for executing any custom logic needed to be used for finding the right store to redirect to. For example, I have a current use-case where we find the user's preferred locale through a series of checks, and if we can determine their preferred locale we will use the closest store that matches it. I wanted to be able to execute this without modifying this module directly, as that is a unique requirement for us.

### Screenshots of Visual Changes before/after (if There Are Any)
<!-- if you made any changes in the UI layer please provide before/after screenshots -->

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [x] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [ ] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [ ] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance
<!-- Please get familiar with following info -->

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/DivanteLtd/vue-storefront?pullRequest=4955) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/DivanteLtd/vue-storefront?pullRequest=4955) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/DivanteLtd/vue-storefront?pullRequest=4955) it.</sub>

rain2o commented:
I'm going to update this today to include a config option for the `blacklist`, as some situations might want specific paths to explicitly not have a store code. We have a few situations that match this requirement, so I imagine others will as well.

Fifciu commented:
Amazing work! It looks really cool. However, could you please write unit tests for this module?

rain2o commented:
@Fifciu Sure thing, I'll add those tests as soon as I can.

gibkigonzo commented on src/modules/default-store-redirect/server.ts:
```diff
@@ -0,0 +1,44 @@
+      `^((?!\\/(${storeCodesStr})(\\/|$))(?!\\/(${blacklistStr})$))\\/?.*((?<=\\.html)|(?<!\\.[a-zA-Z0-9]*))$`
+    )
+
+    app.get(hasStoreCode, (req, res) => {
```
`blacklist` config should not exist. It's like a solution for a problem that we want to create here :thinking: We can move this check here https://github.com/DivanteLtd/vue-storefront/blob/master/core/scripts/server.ts#L167 and return from `app.get('*', (req, res, next) => {` if route is redirected. Or we can add middleware before that https://github.com/DivanteLtd/vue-storefront/blob/master/core/scripts/server.ts#L161 (it needs to be tested, but we should have in that point only routes of our app, not webpack files)

rain2o replied:
@gibkigonzo Yes, the blacklist I originally copied over from the source I used as a starting point for this module. I also intended to make it a config field because we had a situation where we have some specific routes that needed to be non-localized. But after reviewing this, it didn't make sense to do this and we will be removing this need. I'll remove the blacklist here as well.

Are you suggesting I move the blacklist check into server.ts somewhere? I don't fully understand what your suggestion is.

rain2o commented on docs/guide/basics/configuration.md:
```diff
@@ -161,6 +161,11 @@ This both option is used when you don't want re-attempting task of just X number
 
 This option is used only in the [Multistore setup](../integrations/multistore.md). By default it's `''` but if you're running, for example, a multi-instance Vue Storefront setup and the current instance shall be connected to the `en` store on the backend, please just set it so. This config variable is referenced in the [core/lib/multistore.ts](https://github.com/DivanteLtd/vue-storefront/blob/mas ...
 
+If you would like requests without a storeCode to be redirected to the `defaultStoreCode`, you can optionally set the following to `true`.
```
@gibkigonzo I understand what you are explaining, but I don't understand what the requested change is. I only added the instructions to optionally force a redirect to the default store code, as opposed to the current behavior of VSF which is to load the default store without adding the store code to the url.

rain2o commented on src/modules/default-store-redirect/server.ts:
```diff
@@ -0,0 +1,44 @@
+    const blacklistStr = blacklist.join('|')
+    const storeCodesStr = storeCodes.join('|')
+    const hasStoreCode = new RegExp(
+      `^((?!\\/(${storeCodesStr})(\\/|$))(?!\\/(${blacklistStr})$))\\/?.*((?<=\\.html)|(?<!\\.[a-zA-Z0-9]*))$`
```
I will review this. I originally used this as I was implementing the solution someone else provided as an external module. I think I remember trying `storeCodeFromRoute` instead, but I had some issues with it. I will try again and see if I can make it work.

rain2o commented on docs/guide/basics/configuration.md:
````diff
@@ -161,6 +161,11 @@ This both option is used when you don't want re-attempting task of just X number
 
+If you would like requests without a storeCode to be redirected to the `defaultStoreCode`, you can optionally set the following to `true`.
+```json
+"appendDefaultStoreCode": true,
````
@gibkigonzo This was a mistake on my part. I originally added that, and then later removed it and forgot to remove it from the docs.

Thank you for the info about the multi store refactor, this is very exciting! We have been using some custom workarounds on our shop to accomplish this type of behavior because we couldn't get it working with a different url key than the store code.

rain2o commented:
@gibkigonzo Sorry for the confused messages yesterday. I spent a couple months away from this, and was having trouble getting back into the right focus. I believe I now understand most of your requests better. 

I updated my branch so it has your refactor of the multistore. I then removed this module and moved the logic into server.ts as you suggested. However, if I remove the regex and instead use `storeCodeFromRoute`, then when running `yarn dev` I see the same errors described in issue #4532 . It seems it is trying to pull in some things before they're ready, and I'm not too sure how to move forward with this approach.

bloodf commented:
This is closed due to inactivity.

If you want your PR to be re-opened, please update your files, check for any conflicts.

If your PR is for Vue Storefront 1 target the new repository at https://github.com/vuestorefront/vue-storefront-1

rain2o commented:
@bloodf This ticket was inactive because I have been waiting for someone to respond to my last comment. I tried following the approach suggested but ran into some issues and was waiting for some feedback from the vsf team. I would like to update this and submit it in the future, but need some help resolving the problem mentioned before.
